File: tell-dont-ask-kata-ruby/models/Product.rb
require_relative 'OrderItem'

class Product

    attr_accessor :name, :price, :category

    def initialize(name: 'Product Name', price: 1.25, category: 'Category')
        @name = name
        @price = price
        @category = category
    end

    def get_unitary_tax
        return ((@price / 100.0) * category.tax_percentage).round(2)
    end

    def get_unitary_taxed_amount
        return @price + get_unitary_tax.round(2)
    end

    def get_taxed_amount(quantity:)
        return (get_unitary_taxed_amount * quantity).round(2)
    end

    def construct_order_item(item_quantity:)
        order_item = OrderItem.new(
            :product => self,
            :quantity => item_quantity,
            :tax => get_unitary_tax * item_quantity,
            :taxed_amount => get_taxed_amount(:quantity => item_quantity)
        )

        return order_item
    end
end

File: tell-dont-ask-kata-csharp/TellDontAskKata/domain/Product.cs
using System;

public class Product
{
    private string name;
    private decimal price;
    private Category category;

    public string GetName()
    {
        return name;
    }

    public void SetName(string name)
    {
        this.name = name;
    }

    public decimal GetPrice()
    {
        return price;
    }

    public void SetPrice(decimal price)
    {
        this.price = price;
    }

    public void SetCategory(Category category)
    {
        this.category = category;
    }

    public decimal GetUnitaryTax()
    {
        return Math.Round(price / 100 * category.GetTaxPercentage(), 2, MidpointRounding.AwayFromZero);
    }

    public decimal GetUnitaryTaxedAmount()
    {
        return Math.Round(price + GetUnitaryTax(),2, MidpointRounding.AwayFromZero);
    }

    public decimal GetTaxedAmount(int quantity)
    {
        return Math.Round(GetUnitaryTaxedAmount() * Convert.ToDecimal(quantity), 2, MidpointRounding.AwayFromZero);
    }

    public OrderItem ConstructOrderItem(int quantity)
    {
        OrderItem orderItem = new OrderItem();
        orderItem.SetProduct(this);
        orderItem.SetQuantity(quantity);
        orderItem.SetTax(GetUnitaryTax() * quantity);
        orderItem.SetTaxedAmount(GetTaxedAmount(quantity));
        return orderItem;
    }
}